### ISSUE TITLE
can't plot filtered DEGs

### ISSUE THREAD

**Issue by dm8000** [scverse/scanpy#1941]
- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the latest version of scanpy.
- [ ] (optional) I have confirmed this bug exists on the master branch of scanpy.

---
Hello
This block seemed to be running fine a few weeks ago. I tested on the same AnnData file. I couldn't find other issue related to this and not sure how to work around it.

Code:
```
#use this
sc.pl.rank_genes_groups_heatmap(adata, n_genes=200, standard_scale='var', key='rank_genes_groups_filtered', swap_axes=True)
sc.pl.correlation_matrix(adata, 'leiden_0.8', figsize=(14,6))
sc.pl.rank_genes_groups_stacked_violin(adata, groupby = 'leiden_0.8', n_genes=2, key='rank_genes_groups_filtered')
```

Error:

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-16-30381f660d76> in <module>
      1 #use this
----> 2 sc.pl.rank_genes_groups_heatmap(adata, n_genes=200, standard_scale='var', key='rank_genes_groups_filtered', swap_axes=True)
      3 sc.pl.correlation_matrix(adata, 'leiden_0.8', figsize=(14,6))
      4 sc.pl.rank_genes_groups_stacked_violin(adata, groupby = 'leiden_0.8', n_genes=2, key='rank_genes_groups_filtered')

/data04/projects04/MarianaBoroni/lbbc_members/lib/conda_envs/diogoamb/lib/python3.9/site-packages/scanpy/plotting/_tools/__init__.py in rank_genes_groups_heatmap(adata, groups, n_genes, groupby, gene_symbols, var_names, min_logfoldchange, key, show, save, **kwds)
    659     tl.dendrogram
    660     """
--> 661     return _rank_genes_groups_plot(
    662         adata,
    663         plot_type='heatmap',

/data04/projects04/MarianaBoroni/lbbc_members/lib/conda_envs/diogoamb/lib/python3.9/site-packages/scanpy/plotting/_tools/__init__.py in _rank_genes_groups_plot(adata, plot_type, groups, n_genes, groupby, values_to_plot, var_names, min_logfoldchange, key, show, save, return_fig, gene_symbols, **kwds)
    578         from .._anndata import heatmap
    579 
--> 580         return heatmap(
    581             adata,
    582             var_names,

/data04/projects04/MarianaBoroni/lbbc_members/lib/conda_envs/diogoamb/lib/python3.9/site-packages/scanpy/plotting/_anndata.py in heatmap(adata, var_names, groupby, use_raw, log, num_categories, dendrogram, gene_symbols, var_group_positions, var_group_labels, var_group_rotation, layer, standard_scale, swap_axes, show_gene_labels, show, save, figsize, vmin, vmax, vcenter, norm, **kwds)
   1022     )
   1023 
-> 1024     categories, obs_tidy = _prepare_dataframe(
   1025         adata,
   1026         var_names,

/data04/projects04/MarianaBoroni/lbbc_members/lib/conda_envs/diogoamb/lib/python3.9/site-packages/scanpy/plotting/_anndata.py in _prepare_dataframe(adata, var_names, groupby, use_raw, log, num_categories, layer, gene_symbols)
   1917         groupby.remove(groupby_index)
   1918     keys = list(groupby) + list(np.unique(var_names))
-> 1919     obs_tidy = get.obs_df(
   1920         adata, keys=keys, layer=layer, use_raw=use_raw, gene_symbols=gene_symbols
   1921     )

/data04/projects04/MarianaBoroni/lbbc_members/lib/conda_envs/diogoamb/lib/python3.9/site-packages/scanpy/get/get.py in obs_df(adata, keys, obsm_keys, layer, gene_symbols, use_raw)
    270         alias_index = None
    271 
--> 272     obs_cols, var_idx_keys, var_symbols = _check_indices(
    273         adata.obs,
    274         var.index,

/data04/projects04/MarianaBoroni/lbbc_members/lib/conda_envs/diogoamb/lib/python3.9/site-packages/scanpy/get/get.py in _check_indices(dim_df, alt_index, dim, keys, alias_index, use_raw)
    165             not_found.append(key)
    166     if len(not_found) > 0:
--> 167         raise KeyError(
    168             f"Could not find keys '{not_found}' in columns of `adata.{dim}` or in"
    169             f" {alt_repr}.{alt_search_repr}."

KeyError: "Could not find keys '['nan']' in columns of `adata.obs` or in adata.raw.var_names."
```

**Comment by Koncopd** [comment]
What is in `adata.uns['rank_genes_groups_filtered']['params']['groupby']`?

**Comment by dm8000** [comment]
> What is in `adata.uns['rank_genes_groups_filtered']['params']['groupby']`?

the output is `'leiden_0.8'`

**Comment by Koncopd** [comment]
And some names are `nan` in `adata.uns['rank_genes_groups_filtered']['names']` (due to filtering), aren't they?

**Comment by Koncopd** [comment]
I don't think it works for filtered `rank_genes_groups`.
You can call it on unfiltered and specify `min_logfoldchange` directly in the plotting function if you only need this.
Or if you really need the full filtering, then a workaround for now can be gathering the proper names of the genes (omitting `nan`s) from `adata.uns['rank_genes_groups_filtered']['names']` with the groups names as keys and passing it to `sc.pl.rank_genes_groups_heatmap(..., var_names=your_dict)`.
I can write an example and check if you need it.

**Comment by dm8000** [comment]
yes, apparently the problem is associated with these 'nan' values. I assumed this has to do with a package update, because it was working before. I rolled back to scanpy 1.7.1 and this block is working fine. The 'nan's are just omitted in the plots.

I appreciate you offered to give a code example as my python skills are limited. But I figured a workaround so you don't have to bother with it. 

I won't close the issue yet because I assume this behavior was not intended with the update.
